src/webserver.py

```diff
@@ -0,0 +1,156 @@
+            config_entries.append(f"{setting} = {setting_value}")
+        return "\n".join(config_entries)
+
+    @property
+    def _config_path(self) -> pathlib.Path:
+        """Gets the path to the Gunicorn configuration file.
+
+        Returns:
+            The path to the web server configuration file.
+        """
+        return self._charm_state.base_dir / "gunicorn.conf.py"
+
+    @property
+    def command(self) -> list[str]:
+        """Get the command to start the Gunicorn web server.
+
+        Returns:
+            The command to start the Gunicorn web server.
+        """
+        return [
+            "python3",
+            "-m",
+            "gunicorn",
+            "-c",
+            str(self._config_path),
+            self._charm_state.flask_wsgi_app_path,
+        ]
+
+    @property
+    def _check_config_command(self) -> list[str]:
+        """Returns the command to check the Gunicorn configuration.
+
+        Returns:
+            The command to check the Gunicorn configuration.
+        """
+        return self.command + ["--check-config"]
+
+    @property
+    def _reload_signal(self) -> signal.Signals:
+        """Get the signal used to reload the Gunicorn web server.
+
+        Returns:
+            The signal used to reload the Gunicorn web server.
+        """
+        return signal.SIGHUP
+
+    def _exec(self, command: list[str]) -> ExecResult:
+        """Execute a command inside the Flask workload container.
+
+        The command will be executed with user flask group flask inside the container.
+
+        Args:
+            command: A list of strings representing the command to be executed.
+
+        Returns:
+            ExecResult: An `ExecResult` object representing the result of the command execution.
+        """
+        container = self._flask_container
+        exec_process = container.exec(command, user="flask", group="flask")
+        try:
+            stdout, stderr = exec_process.wait_output()
+            return ExecResult(0, typing.cast(str, stdout), typing.cast(str, stderr))
+        except ExecError as exc:
+            return ExecResult(
+                exc.exit_code, typing.cast(str, exc.stdout), typing.cast(str, exc.stderr)
+            )
+
+    def update_config(self, is_webserver_running: bool) -> None:
+        """Update and apply the configuration file of the web server.
+
+        Args:
+            is_webserver_running: Indicates if the web server container is currently running.
+
+        Raises:
+            WebserverConfigInvalidError: If the web server configuration check fails.
+        """
+        webserver_config_path = str(self._config_path)
+        try:
+            current_webserver_config = self._flask_container.pull(webserver_config_path)
+        except PathError:
+            current_webserver_config = None
+        self._flask_container.push(webserver_config_path, self._config)
+        if current_webserver_config == self._config:
+            return
```

Comment:
This will happen every time there's a config-changed event, which means it will happen quite frequently. I feel it might be a little bit redundant to notify the user that configuration hasn't been changed. The flask charm prints a log message when the configuration changed, do you think that's good enough?

Reply:
I see, you're right, it would be redundant.